File: sdks/py/src/jeffs_brain_memory/memory/heuristic.py
# SPDX-License-Identifier: Apache-2.0
"""Heuristic memory file management (reinforcement + listing)."""

from __future__ import annotations

import re
from datetime import datetime, timezone
from typing import TYPE_CHECKING

from ._memstore import ListOpts, NotFoundError
from .paths import (
    base_name,
    memory_global_index,
    memory_global_prefix,
    memory_global_topic,
    memory_project_index,
    memory_project_prefix,
    memory_project_topic,
)
from .reflect import Heuristic
from .store import parse_frontmatter
from .types import TopicFile

if TYPE_CHECKING:
    from .manager import MemoryManager

STOP_WORDS = {
    "the",
    "a",
    "an",
    "in",
    "on",
    "for",
    "to",
    "of",
    "with",
    "when",
    "and",
    "or",
    "but",
    "is",
    "are",
    "was",
    "were",
    "be",
    "not",
    "do",
}

_NON_ALPHA_NUM = re.compile(r"[^a-z0-9 ]")


class HeuristicSummary:
    def __init__(
        self,
        name: str,
        path: str,
        confidence: str,
        tags: list[str],
        scope: str,
        is_anti: bool,
    ) -> None:
        self.name = name
        self.path = path
        self.confidence = confidence
        self.tags = tags
        self.scope = scope
        self.is_anti = is_anti


def apply_heuristics(
    mem: "MemoryManager", project_slug: str, heuristics: list[Heuristic]
) -> None:
    project_entries: list[str] = []
    global_entries: list[str] = []
    writes: list[tuple[str, bytes]] = []

    for h in heuristics:
        if not h.rule:
            continue
        if h.scope == "global":
            prefix = memory_global_prefix()
        else:
            prefix = memory_project_prefix(project_slug)
        existing_path, existing_content, found = _find_existing_heuristic(
            mem, h, prefix
        )
        if found:
            content = merge_heuristic(existing_content, h)
            path = existing_path
        else:
            content = build_heuristic_content(h)
            fname_no_ext = heuristic_filename(h)[: -len(".md")]
            if h.scope == "global":
                path = memory_global_topic(fname_no_ext)
            else:
                path = memory_project_topic(project_slug, fname_no_ext)

        writes.append((path, content.encode("utf-8")))

        entry = f"- [heuristic] {base_name(path)}: {truncate(h.rule, 100)}"
        if h.scope == "global":
            global_entries.append(entry)
        else:
            project_entries.append(entry)

    if not writes:
        return

    def _run(b) -> None:
        for p, c in writes:
            b.write(p, c)
        if project_entries:
            _append_index_entries(b, memory_project_index(project_slug), project_entries)
        if global_entries:
            _append_index_entries(b, memory_global_index(), global_entries)

    mem.store.batch(_run)


def _find_existing_heuristic(
    mem: "MemoryManager", h: Heuristic, prefix: str
) -> tuple[str, str, bool]:
    try:
        entries = mem.store.list(prefix, ListOpts(include_generated=True))
    except NotFoundError:
        return "", "", False

    candidate_words = significant_words(heuristic_filename(h))
    for entry in entries:
        if entry.is_dir:
            continue
        name = base_name(entry.path)
        if not name.endswith(".md") or name.lower() == "memory.md":
            continue
        try:
            data = mem.store.read(entry.path)
        except NotFoundError:
            continue
        fm, _ = parse_frontmatter(data.decode("utf-8"))
        if not has_tag(fm.tags, "heuristic"):
            continue
        if not has_tag(fm.tags, h.category):
            continue
        existing_words = significant_words(name)
        if jaccard_similarity(candidate_words, existing_words) > 0.5:
            return entry.path, data.decode("utf-8"), True
    return "", "", False


def heuristic_filename(h: Heuristic) -> str:
    words = significant_words(h.rule)
    slug = h.category or "general"
    prefix = "heuristic-anti" if h.anti_pattern else "heuristic"
    limit = min(2, len(words))
    parts = [prefix, slug] + words[:limit]
    return "-".join(parts) + ".md"


def significant_words(text: str) -> list[str]:
    cleaned = _NON_ALPHA_NUM.sub(" ", text.lower())
    result: list[str] = []
    for w in cleaned.split():
        if w not in STOP_WORDS and len(w) > 1:
            result.append(w)
    return result


def build_heuristic_content(h: Heuristic) -> str:
    now = datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")
    heading = first_n_words(h.rule, 5)
    lines = ["---"]
    lines.append(f'name: "{capitalise(h.category)}: {heading}"')
    lines.append(f'description: "{truncate(h.rule, 100)}"')
    lines.append("type: feedback")
    lines.append(f"created: {now}")
    lines.append(f"modified: {now}")
    lines.append(f"confidence: {h.confidence}")
    lines.append("source: reflection")
    lines.append("tags:")
    lines.append("  - heuristic")
    lines.append(f"  - {h.category}")
    if h.anti_pattern:
        lines.append("  - anti-pattern")
    lines.append("---")
    lines.append("")
    if h.anti_pattern:
        lines.append(f"## Anti-pattern: {heading}")
        lines.append("")
        lines.append(f"**Don't:** {h.rule}")
        lines.append("")
        alt = extract_alternative(h.rule)
        if alt:
            lines.append(f"**Instead:** {alt}")
            lines.append("")
        lines.append("**Why:** Observed during reflection")
        lines.append("")
        lines.append(f"**Confidence:** {h.confidence} (1 observation)")
    else:
        lines.append(f"## {heading}")
        lines.append("")
        lines.append(h.rule)
        lines.append("")
        if h.context:
            lines.append(f"**Context:** {h.context}")
            lines.append("")
        lines.append("**Why:** Observed during reflection")
        lines.append("")
        lines.append(f"**Confidence:** {h.confidence} (1 observation)")
    return "\n".join(lines) + "\n"


def merge_heuristic(existing_content: str, h: Heuristic) -> str:
    fm, body = parse_frontmatter(existing_content)
    count = count_sections(body)
    new_confidence = confidence_from_observations(count + 1)

    now = datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")
    heading = first_n_words(h.rule, 5)
    section_lines: list[str] = [""]
    if h.anti_pattern:
        section_lines.append(f"## Anti-pattern: {heading}")
        section_lines.append("")
        section_lines.append(f"**Don't:** {h.rule}")
        section_lines.append("")
        section_lines.append("**Why:** Observed during reflection")
        section_lines.append("")
        section_lines.append(f"**Confidence:** {new_confidence} ({count + 1} observations)")
    else:
        section_lines.append(f"## {heading}")
        section_lines.append("")
        section_lines.append(h.rule)
        section_lines.append("")
        if h.context:
            section_lines.append(f"**Context:** {h.context}")
            section_lines.append("")
        section_lines.append("**Why:** Observed during reflection")
        section_lines.append("")
        section_lines.append(f"**Confidence:** {new_confidence} ({count + 1} observations)")

    lines = ["---"]
    if fm.name:
        lines.append(f'name: "{fm.name}"')
    if fm.description:
        lines.append(f'description: "{fm.description}"')
    if fm.type:
        lines.append(f"type: {fm.type}")
    if fm.created:
        lines.append(f"created: {fm.created}")
    lines.append(f"modified: {now}")
    lines.append(f"confidence: {new_confidence}")
    if fm.source:
        lines.append(f"source: {fm.source}")
    if fm.tags:
        lines.append("tags:")
        for tag in fm.tags:
            lines.append(f"  - {tag}")
    lines.append("---")
    lines.append("")
    lines.append(body.strip())
    lines.append("")
    lines.extend(section_lines)
    return "\n".join(lines)


def confidence_from_observations(count: int) -> str:
    if count >= 4:
        return "high"
    if count >= 2:
        return "medium"
    return "low"


def has_tag(tags: list[str], tag: str) -> bool:
    target = tag.lower()
    return any(t.lower() == target for t in tags)


def jaccard_similarity(a: list[str], b: list[str]) -> float:
    if not a and not b:
        return 1.0
    set_a = set(a)
    set_b = set(b)
    intersection = len(set_a & set_b)
    union = len(set_a | set_b)
    if union == 0:
        return 0.0
    return intersection / union


def count_sections(body: str) -> int:
    count = 0
    for line in body.split("\n"):
        if line.strip().startswith("## "):
            count += 1
    return count


def first_n_words(text: str, n: int) -> str:
    words = text.split()
    if len(words) > n:
        words = words[:n]
    return " ".join(words)


def truncate(text: str, max_len: int) -> str:
    if len(text) <= max_len:
        return text
    if max_len < 4:
        return text[:max_len]
    return text[: max_len - 3] + "..."


def capitalise(s: str) -> str:
    if not s:
        return s
    return s[0].upper() + s[1:]


def extract_alternative(rule: str) -> str:
    lower = rule.lower()
    for marker in ("instead ", "use ", "prefer "):
        idx = lower.find(marker)
        if idx >= 0:
            return rule[idx:].strip()
    return ""


def _append_index_entries(b, index_path: str, entries: list[str]) -> None:
    content = ""
    try:
        existing = b.read(index_path)
        content = existing.decode("utf-8").strip()
    except NotFoundError:
        content = ""
    for entry in entries:
        entry = entry.strip()
        if not entry:
            continue
        if entry in content:
            continue
        if content:
            content += "\n"
        content += entry
    b.write(index_path, (content + "\n").encode("utf-8"))


def list_heuristics_in(mem: "MemoryManager", project_path: str) -> list[HeuristicSummary]:
    summaries: list[HeuristicSummary] = []
    for t in mem.list_project_topics(project_path):
        if has_tag(t.tags, "heuristic"):
            summaries.append(
                HeuristicSummary(
                    name=t.name,
                    path=t.path,
                    confidence=t.confidence,
                    tags=list(t.tags),
                    scope="project",
                    is_anti=has_tag(t.tags, "anti-pattern"),
                )
            )
    for t in mem.list_global_topics():
        if has_tag(t.tags, "heuristic"):
            summaries.append(
                HeuristicSummary(
                    name=t.name,
                    path=t.path,
                    confidence=t.confidence,
                    tags=list(t.tags),
                    scope="global",
                    is_anti=has_tag(t.tags, "anti-pattern"),
                )
            )
    return summaries

File: go/memory/heuristic.go
// SPDX-License-Identifier: Apache-2.0

package memory

import (
	"context"
	"fmt"
	"regexp"
	"strings"
	"time"

	"github.com/jeffs-brain/memory/go/brain"
)

// HeuristicSummary is a lightweight view of a heuristic memory file,
// used by callers that list or display heuristics.
type HeuristicSummary struct {
	Name       string
	Path       brain.Path
	Confidence string
	Tags       []string
	Scope      string
	IsAnti     bool
}

// stopWords are filtered out when extracting significant words from a rule.
var stopWords = map[string]bool{
	"the": true, "a": true, "an": true, "in": true, "on": true,
	"for": true, "to": true, "of": true, "with": true, "when": true,
	"and": true, "or": true, "but": true, "is": true, "are": true,
	"was": true, "were": true, "be": true, "not": true, "do": true,
}

// nonAlphaNum matches any character that is not a letter, digit, or
// space.
var nonAlphaNum = regexp.MustCompile(`[^a-z0-9 ]`)

// ApplyHeuristics writes or updates heuristic memory files based on
// reflection output. All writes happen in a single batch so heuristics
// land as one coherent commit.
func (m *Memory) ApplyHeuristics(ctx context.Context, projectSlug string, heuristics []Heuristic) error {
	var projectEntries, globalEntries []string

	type pending struct {
		path    brain.Path
		content []byte
	}
	var writes []pending

	for _, h := range heuristics {
		if h.Rule == "" {
			continue
		}

		var prefix brain.Path
		if h.Scope == "global" {
			prefix = brain.MemoryGlobalPrefix()
		} else {
			prefix = brain.MemoryProjectPrefix(projectSlug)
		}

		existingPath, existingContent, found := m.findExistingHeuristic(ctx, h, prefix)

		var path brain.Path
		var content string

		if found {
			content = mergeHeuristic(existingContent, h)
			path = existingPath
		} else {
			content = buildHeuristicContent(h)
			if h.Scope == "global" {
				path = brain.MemoryGlobalTopic(strings.TrimSuffix(heuristicFilename(h), ".md"))
			} else {
				path = brain.MemoryProjectTopic(projectSlug, strings.TrimSuffix(heuristicFilename(h), ".md"))
			}
		}

		writes = append(writes, pending{path: path, content: []byte(content)})

		entry := fmt.Sprintf("- [heuristic] %s: %s", baseName(string(path)), truncate(h.Rule, 100))
		if h.Scope == "global" {
			globalEntries = append(globalEntries, entry)
		} else {
			projectEntries = append(projectEntries, entry)
		}
	}

	if len(writes) == 0 {
		return nil
	}

	return m.store.Batch(ctx, brain.BatchOptions{Reason: "reflect"}, func(b brain.Batch) error {
		for _, w := range writes {
			if err := b.Write(ctx, w.path, w.content); err != nil {
				return err
			}
		}
		if len(projectEntries) > 0 {
			if err := m.appendIndexEntries(ctx, b, brain.MemoryProjectIndex(projectSlug), projectEntries); err != nil {
				return err
			}
		}
		if len(globalEntries) > 0 {
			if err := m.appendIndexEntries(ctx, b, brain.MemoryGlobalIndex(), globalEntries); err != nil {
				return err
			}
		}
		return nil
	})
}

// heuristicFilename generates a filename from the heuristic's category
// and the first two significant words of the rule.
func heuristicFilename(h Heuristic) string {
	words := significantWords(h.Rule)
	slug := h.Category
	if slug == "" {
		slug = "general"
	}

	prefix := "heuristic"
	if h.AntiPattern {
		prefix = "heuristic-anti"
	}

	limit := 2
	if len(words) < limit {
		limit = len(words)
	}

	parts := []string{prefix, slug}
	parts = append(parts, words[:limit]...)

	return strings.Join(parts, "-") + ".md"
}

// significantWords extracts lowercase words from text, filtering out
// stop words and non-alphanumeric characters.
func significantWords(text string) []string {
	cleaned := nonAlphaNum.ReplaceAllString(strings.ToLower(text), " ")
	raw := strings.Fields(cleaned)

	var result []string
	for _, w := range raw {
		if !stopWords[w] && len(w) > 1 {
			result = append(result, w)
		}
	}
	return result
}

// findExistingHeuristic searches under the given prefix for an existing
// heuristic file matching the given heuristic.
func (m *Memory) findExistingHeuristic(ctx context.Context, h Heuristic, prefix brain.Path) (path brain.Path, content string, found bool) {
	entries, err := m.store.List(ctx, prefix, brain.ListOpts{IncludeGenerated: true})
	if err != nil {
		return "", "", false
	}

	candidateWords := significantWords(heuristicFilename(h))

	for _, entry := range entries {
		if entry.IsDir {
			continue
		}
		name := baseName(string(entry.Path))
		if !strings.HasSuffix(name, ".md") {
			continue
		}
		if strings.EqualFold(name, "MEMORY.md") {
			continue
		}

		data, err := m.store.Read(ctx, entry.Path)
		if err != nil {
			continue
		}

		fm, _ := ParseFrontmatter(string(data))

		if !hasTag(fm.Tags, "heuristic") {
			continue
		}
		if !hasTag(fm.Tags, h.Category) {
			continue
		}

		existingWords := significantWords(name)
		if jaccardSimilarity(candidateWords, existingWords) > 0.5 {
			return entry.Path, string(data), true
		}
	}

	return "", "", false
}

// buildHeuristicContent produces a complete heuristic memory file with
// YAML frontmatter and structured body.
func buildHeuristicContent(h Heuristic) string {
	now := time.Now().UTC().Format(time.RFC3339)
	heading := firstNWords(h.Rule, 5)

	var b strings.Builder

	b.WriteString("---\n")
	b.WriteString(fmt.Sprintf("name: \"%s: %s\"\n", capitalise(h.Category), heading))
	b.WriteString(fmt.Sprintf("description: \"%s\"\n", truncate(h.Rule, 100)))
	b.WriteString("type: feedback\n")
	b.WriteString(fmt.Sprintf("created: %s\n", now))
	b.WriteString(fmt.Sprintf("modified: %s\n", now))
	b.WriteString(fmt.Sprintf("confidence: %s\n", h.Confidence))
	b.WriteString("source: reflection\n")
	b.WriteString("tags:\n")
	b.WriteString("  - heuristic\n")
	b.WriteString(fmt.Sprintf("  - %s\n", h.Category))
	if h.AntiPattern {
		b.WriteString("  - anti-pattern\n")
	}
	b.WriteString("---\n\n")

	if h.AntiPattern {
		b.WriteString(fmt.Sprintf("## Anti-pattern: %s\n\n", heading))
		b.WriteString(fmt.Sprintf("**Don't:** %s\n\n", h.Rule))
		if alt := extractAlternative(h.Rule); alt != "" {
			b.WriteString(fmt.Sprintf("**Instead:** %s\n\n", alt))
		}
		b.WriteString("**Why:** Observed during reflection\n\n")
		b.WriteString(fmt.Sprintf("**Confidence:** %s (1 observation)\n", h.Confidence))
	} else {
		b.WriteString(fmt.Sprintf("## %s\n\n", heading))
		b.WriteString(fmt.Sprintf("%s\n\n", h.Rule))
		if h.Context != "" {
			b.WriteString(fmt.Sprintf("**Context:** %s\n\n", h.Context))
		}
		b.WriteString("**Why:** Observed during reflection\n\n")
		b.WriteString(fmt.Sprintf("**Confidence:** %s (1 observation)\n", h.Confidence))
	}

	return b.String()
}

// mergeHeuristic adds a new observation section to an existing heuristic
// file.
func mergeHeuristic(existingContent string, h Heuristic) string {
	fm, body := ParseFrontmatter(existingContent)
	count := countSections(body)
	newConfidence := confidenceFromObservations(count + 1)

	now := time.Now().UTC().Format(time.RFC3339)
	fm.Modified = now
	fm.Confidence = newConfidence

	heading := firstNWords(h.Rule, 5)
	var section strings.Builder

	if h.AntiPattern {
		section.WriteString(fmt.Sprintf("\n\n## Anti-pattern: %s\n\n", heading))
		section.WriteString(fmt.Sprintf("**Don't:** %s\n\n", h.Rule))
		section.WriteString("**Why:** Observed during reflection\n\n")
		section.WriteString(fmt.Sprintf("**Confidence:** %s (%d observations)\n", newConfidence, count+1))
	} else {
		section.WriteString(fmt.Sprintf("\n\n## %s\n\n", heading))
		section.WriteString(fmt.Sprintf("%s\n\n", h.Rule))
		if h.Context != "" {
			section.WriteString(fmt.Sprintf("**Context:** %s\n\n", h.Context))
		}
		section.WriteString("**Why:** Observed during reflection\n\n")
		section.WriteString(fmt.Sprintf("**Confidence:** %s (%d observations)\n", newConfidence, count+1))
	}

	var b strings.Builder
	b.WriteString("---\n")
	if fm.Name != "" {
		b.WriteString(fmt.Sprintf("name: \"%s\"\n", fm.Name))
	}
	if fm.Description != "" {
		b.WriteString(fmt.Sprintf("description: \"%s\"\n", fm.Description))
	}
	if fm.Type != "" {
		b.WriteString(fmt.Sprintf("type: %s\n", fm.Type))
	}
	if fm.Created != "" {
		b.WriteString(fmt.Sprintf("created: %s\n", fm.Created))
	}
	b.WriteString(fmt.Sprintf("modified: %s\n", now))
	b.WriteString(fmt.Sprintf("confidence: %s\n", newConfidence))
	if fm.Source != "" {
		b.WriteString(fmt.Sprintf("source: %s\n", fm.Source))
	}
	if len(fm.Tags) > 0 {
		b.WriteString("tags:\n")
		for _, tag := range fm.Tags {
			b.WriteString(fmt.Sprintf("  - %s\n", tag))
		}
	}
	b.WriteString("---\n\n")

	b.WriteString(strings.TrimSpace(body))
	b.WriteString(section.String())

	return b.String()
}

// confidenceFromObservations returns a confidence level based on the
// number of times a heuristic has been observed.
func confidenceFromObservations(count int) string {
	switch {
	case count >= 4:
		return "high"
	case count >= 2:
		return "medium"
	default:
		return "low"
	}
}

// ListHeuristicsIn returns all heuristic memory files from both project
// and global memory for the given project path.
func (m *Memory) ListHeuristicsIn(ctx context.Context, projectPath string) []HeuristicSummary {
	var summaries []HeuristicSummary

	projectTopics, _ := m.ListProjectTopics(ctx, projectPath)
	for _, t := range projectTopics {
		if hasTag(t.Tags, "heuristic") {
			summaries = append(summaries, HeuristicSummary{
				Name:       t.Name,
				Path:       t.Path,
				Confidence: t.Confidence,
				Tags:       t.Tags,
				Scope:      "project",
				IsAnti:     hasTag(t.Tags, "anti-pattern"),
			})
		}
	}

	globalTopics, _ := m.ListGlobalTopics(ctx)
	for _, t := range globalTopics {
		if hasTag(t.Tags, "heuristic") {
			summaries = append(summaries, HeuristicSummary{
				Name:       t.Name,
				Path:       t.Path,
				Confidence: t.Confidence,
				Tags:       t.Tags,
				Scope:      "global",
				IsAnti:     hasTag(t.Tags, "anti-pattern"),
			})
		}
	}

	return summaries
}

// hasTag checks whether a tag slice contains the given tag
// (case-insensitive).
func hasTag(tags []string, tag string) bool {
	target := strings.ToLower(tag)
	for _, t := range tags {
		if strings.ToLower(t) == target {
			return true
		}
	}
	return false
}

// jaccardSimilarity computes the Jaccard index between two word slices.
func jaccardSimilarity(a, b []string) float64 {
	if len(a) == 0 && len(b) == 0 {
		return 1.0
	}

	setA := make(map[string]bool, len(a))
	for _, w := range a {
		setA[w] = true
	}

	setB := make(map[string]bool, len(b))
	for _, w := range b {
		setB[w] = true
	}

	intersection := 0
	for w := range setA {
		if setB[w] {
			intersection++
		}
	}

	union := len(setA)
	for w := range setB {
		if !setA[w] {
			union++
		}
	}

	if union == 0 {
		return 0
	}
	return float64(intersection) / float64(union)
}

// countSections counts the number of ## headings in a markdown body.
func countSections(body string) int {
	count := 0
	for _, line := range strings.Split(body, "\n") {
		if strings.HasPrefix(strings.TrimSpace(line), "## ") {
			count++
		}
	}
	return count
}

// firstNWords returns the first n words of text, joined by spaces.
func firstNWords(text string, n int) string {
	words := strings.Fields(text)
	if len(words) > n {
		words = words[:n]
	}
	return strings.Join(words, " ")
}

// truncate shortens text to maxLen characters, appending "..." if
// truncated.
func truncate(text string, maxLen int) string {
	if len(text) <= maxLen {
		return text
	}
	if maxLen < 4 {
		return text[:maxLen]
	}
	return text[:maxLen-3] + "..."
}

// capitalise returns the string with the first letter upper-cased.
func capitalise(s string) string {
	if s == "" {
		return s
	}
	return strings.ToUpper(s[:1]) + s[1:]
}

// extractAlternative attempts to find an alternative suggestion in a
// rule string.
func extractAlternative(rule string) string {
	lower := strings.ToLower(rule)
	for _, marker := range []string{"instead ", "use ", "prefer "} {
		if idx := strings.Index(lower, marker); idx >= 0 {
			return strings.TrimSpace(rule[idx:])
		}
	}
	return ""
}
